# డేటా తయారీ

[మూల నోట్‌బుక్ మూలం *డేటా సైన్స్: డేటా సైన్స్ పైథాన్ మరియు మెషీన్ లెర్నింగ్ స్టూడియో కోసం మెషీన్ లెర్నింగ్ పరిచయం లీ స్టాట్* నుండి](https://github.com/leestott/intro-Datascience/blob/master/Course%20Materials/4-Cleaning_and_Manipulating-Reference.ipynb)

## `DataFrame` సమాచారం అన్వేషణ

> **అభ్యాస లక్ష్యం:** ఈ ఉపవిభాగం చివరికి, మీరు pandas DataFrames లో నిల్వ చేసిన డేటా గురించి సాధారణ సమాచారాన్ని కనుగొనడంలో సౌకర్యంగా ఉండాలి.

మీరు మీ డేటాను pandas లో లోడ్ చేసిన తర్వాత, అది ఎక్కువగా `DataFrame` లోనే ఉంటుంది. అయితే, మీ `DataFrame` లో 60,000 వరుసలు మరియు 400 కాలమ్స్ ఉంటే, మీరు ఏ విధంగా మీరు పని చేస్తున్నదానిపై అవగాహన పొందడం ప్రారంభిస్తారు? అదృష్టవశాత్తు, pandas కొన్ని సౌకర్యవంతమైన సాధనాలను అందిస్తుంది, అవి `DataFrame` గురించి మొత్తం సమాచారాన్ని త్వరగా చూడటానికి, మొదటి కొన్ని మరియు చివరి కొన్ని వరుసలతో పాటు.

ఈ ఫంక్షనాలిటీని అన్వేషించడానికి, మనం Python scikit-learn లైబ్రరీని దిగుమతి చేసుకుని, ప్రతి డేటా శాస్త్రవేత్త వందల సార్లు చూసిన ఒక ప్రతిష్టాత్మక డేటాసెట్‌ను ఉపయోగిస్తాము: బ్రిటిష్ బయాలజిస్ట్ రోనాల్డ్ ఫిషర్ యొక్క *Iris* డేటా సెట్, అతని 1936 పత్రంలో "టాక్సోనామిక్ సమస్యలలో బహుళ కొలతల ఉపయోగం" లో ఉపయోగించినది:


In [1]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])

### `DataFrame.shape`
మేము `iris_df` అనే వేరియబుల్‌లో ఐరిస్ డేటాసెట్‌ను లోడ్ చేసుకున్నాము. డేటాలోకి దిగేముందు, మన దగ్గర ఉన్న డేటాపాయింట్ల సంఖ్య మరియు డేటాసెట్ యొక్క మొత్తం పరిమాణం తెలుసుకోవడం విలువైనది. మనం వ్యవహరిస్తున్న డేటా పరిమాణాన్ని చూడటం ఉపయోగకరం.


In [2]:
iris_df.shape

(150, 4)

కాబట్టి, మనం 150 వరుసలు మరియు 4 కాలమ్స్ ఉన్న డేటాతో వ్యవహరిస్తున్నాము. ప్రతి వరుస ఒక డేటాపాయింట్‌ను సూచిస్తుంది మరియు ప్రతి కాలమ్ డేటా ఫ్రేమ్‌కు సంబంధించిన ఒక ఫీచర్‌ను సూచిస్తుంది. కాబట్టి, మౌలికంగా, ప్రతి ఒక్కటి 4 ఫీచర్లను కలిగి ఉన్న 150 డేటాపాయింట్లు ఉన్నాయి.

`shape` ఇక్కడ డేటాఫ్రేమ్ యొక్క ఒక లక్షణం మరియు ఫంక్షన్ కాదు, అందుకే ఇది కౌన్సుల జంటతో ముగియదు.


### `DataFrame.columns`
ఇప్పుడు మనం 4 డేటా కాలమ్స్ లోకి వెళదాం. వాటిలో ప్రతి ఒక్కటి ఖచ్చితంగా ఏమి సూచిస్తుంది? `columns` అట్రిబ్యూట్ డేటాఫ్రేమ్ లోని కాలమ్స్ పేరును మనకు ఇస్తుంది.


In [3]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

మనము చూడగలిగినట్లుగా, నాలుగు(4) కాలమ్స్ ఉన్నాయి. `columns` లక్షణం కాలమ్స్ పేరును మనకు తెలియజేస్తుంది మరియు మిగతా ఏమీ కాదు. ఈ లక్షణం ఒక డేటాసెట్ లో ఏ లక్షణాలు ఉన్నాయో గుర్తించాలనుకున్నప్పుడు ప్రాముఖ్యత పొందుతుంది.


### `DataFrame.info`
డేటా పరిమాణం(`shape` లక్షణం ద్వారా ఇవ్వబడింది) మరియు ఫీచర్లు లేదా కాలమ్స్ పేర్లు(`columns` లక్షణం ద్వారా ఇవ్వబడింది) dataset గురించి మనకు ఏదో తెలియజేస్తాయి. ఇప్పుడు, మనం dataset లో మరింత లోతుగా వెళ్ళాలనుకుంటున్నాము. `DataFrame.info()` ఫంక్షన్ దీనికి చాలా ఉపయోగకరం.


In [4]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


ఇక్కడి నుండి, మనం కొన్ని పరిశీలనలు చేయవచ్చు:
1. ప్రతి కాలమ్ యొక్క డేటాటైపు: ఈ డేటాసెట్‌లో, అన్ని డేటా 64-బిట్ ఫ్లోటింగ్-పాయింట్ సంఖ్యలుగా నిల్వ చేయబడింది.
2. నాన్-నల్ విలువల సంఖ్య: నల్ విలువలతో వ్యవహరించడం డేటా సిద్ధతలో ఒక ముఖ్యమైన దశ. ఇది నోట్బుక్‌లో తరువాత చర్చించబడుతుంది.


### DataFrame.describe()
మన డేటాసెట్‌లో చాలా సంఖ్యాత్మక డేటా ఉందని అనుకుందాం. సగటు, మధ్య విలువ, క్వార్టైల్‌లు మొదలైన ఒకవిధమైన గణాంక లెక్కింపులు ప్రతి కాలమ్‌పై వ్యక్తిగతంగా చేయవచ్చు. `DataFrame.describe()` ఫంక్షన్ డేటాసెట్‌లోని సంఖ్యాత్మక కాలమ్‌ల గణాంక సారాంశాన్ని అందిస్తుంది.


In [5]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count         150.000000        150.000000         150.000000        150.000000
mean            5.843333          3.057333           3.758000          1.199333
std             0.828066          0.435866           1.765298          0.762238
min             4.300000          2.000000           1.000000          0.100000
25%             5.100000          2.800000           1.600000          0.300000
50%             5.800000          3.000000           4.350000          1.300000
75%             6.400000          3.300000           5.100000          1.800000
max             7.900000          4.400000           6.900000          2.500000

పై అవుట్పుట్ ప్రతి కాలమ్ యొక్క మొత్తం డేటా పాయింట్లు, సగటు, ప్రమాణ విభిన్నత, కనిష్ఠం, తక్కువ క్వార్టైల్(25%), మధ్యస్థానం(50%), పై క్వార్టైల్(75%) మరియు గరిష్ఠ విలువను చూపిస్తుంది.


### `DataFrame.head`
పై అన్ని ఫంక్షన్లు మరియు లక్షణాలతో, మేము డేటాసెట్ యొక్క టాప్ లెవల్ వీక్షణను పొందాము. ఎన్ని డేటా పాయింట్లు ఉన్నాయో, ఎన్ని ఫీచర్లు ఉన్నాయో, ప్రతి ఫీచర్ యొక్క డేటా రకం మరియు ప్రతి ఫీచర్ కోసం ఎన్ని నాన్-నల్ విలువలు ఉన్నాయో మాకు తెలుసు.

ఇప్పుడు డేటాను స్వయంగా చూడాల్సిన సమయం వచ్చింది. మన `DataFrame` యొక్క మొదటి కొన్ని వరుసలు (మొదటి కొన్ని డేటాపాయింట్లు) ఎలా ఉన్నాయో చూద్దాం:


In [6]:
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

ఇక్కడ అవుట్పుట్‌గా, మనం డేటాసెట్‌లో ఐదు(5) ఎంట్రీలను చూడవచ్చు. ఎడమవైపు ఉన్న సూచికను చూస్తే, ఇవి మొదటి ఐదు వరుసలు అని మనం తెలుసుకోవచ్చు.


### వ్యాయామం:

పై ఉదాహరణ నుండి, డిఫాల్ట్‌గా, `DataFrame.head` ఒక `DataFrame` యొక్క మొదటి ఐదు వరుసలను తిరిగి ఇస్తుంది అని స్పష్టంగా తెలుస్తోంది. క్రింది కోడ్ సెల్‌లో, మీరు ఐదు వరుసల కంటే ఎక్కువ వరుసలను ప్రదర్శించడానికి ఒక మార్గాన్ని కనుగొనగలరా?


In [7]:
# Hint: Consult the documentation by using iris_df.head?

### `DataFrame.tail`
డేటాను చూడటానికి మరో మార్గం ప్రారంభం కాకుండా చివర నుండి చూడటం. `DataFrame.head` యొక్క విరుద్ధం `DataFrame.tail`, ఇది ఒక `DataFrame` యొక్క చివరి ఐదు వరుసలను తిరిగి ఇస్తుంది:


In [8]:
iris_df.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

ప్రాక్టీస్‌లో, మీరు ఆర్డర్ చేయబడిన డేటాసెట్‌లలో అవుట్లయర్లను వెతుకుతున్నప్పుడు, `DataFrame` యొక్క మొదటి కొన్ని వరుసలు లేదా చివరి కొన్ని వరుసలను సులభంగా పరిశీలించగలగడం ఉపయోగకరం.

పై కోడ్ ఉదాహరణల సహాయంతో చూపించిన అన్ని ఫంక్షన్లు మరియు లక్షణాలు, డేటాను చూడటానికి మరియు అనుభూతి పొందటానికి సహాయపడతాయి.

> **Takeaway:** DataFrameలోని సమాచారంపై మెటాడేటాను లేదా దాని మొదటి మరియు చివరి కొన్ని విలువలను చూసే క్రమంలోనే, మీరు వ్యవహరిస్తున్న డేటా యొక్క పరిమాణం, ఆకారం మరియు విషయంపై తక్షణమే ఒక ఆలోచన పొందవచ్చు.


### లేని డేటా
లేని డేటాలోకి మనం దిగుదాం. కొన్ని కాలమ్స్‌లో విలువ నిల్వ చేయబడకపోతే లేని డేటా సంభవిస్తుంది.

ఒక ఉదాహరణ తీసుకుందాం: ఎవరో వారి బరువు గురించి జాగ్రత్తగా ఉంటే, సర్వేలో బరువు ఫీల్డ్‌ను నింపకపోవచ్చు. అప్పుడు ఆ వ్యక్తి కోసం బరువు విలువ లేనిది అవుతుంది.

అధికంగా, వాస్తవ ప్రపంచ డేటాసెట్‌లలో, లేని విలువలు సంభవిస్తాయి.

**పాండాస్ లేని డేటాను ఎలా నిర్వహిస్తుంది**


పాండాస్ లేని విలువలను రెండు విధాలుగా నిర్వహిస్తుంది. మీరు ఇంతకు ముందు చూసిన మొదటి విధానం: `NaN`, లేదా నాట్ ఎ నంబర్. ఇది వాస్తవానికి IEEE ఫ్లోటింగ్-పాయింట్ స్పెసిఫికేషన్‌లో భాగమైన ప్రత్యేక విలువ మరియు ఇది కేవలం లేని ఫ్లోటింగ్-పాయింట్ విలువలను సూచించడానికి ఉపయోగించబడుతుంది.

ఫ్లోట్స్ కాకుండా లేని విలువల కోసం, పాండాస్ పైథాన్ `None` ఆబ్జెక్ట్‌ను ఉపయోగిస్తుంది. మీరు రెండు వేర్వేరు రకాల విలువలను ఎదుర్కొంటారని అనిపించవచ్చు, అవి సారాంశంగా అదే విషయాన్ని చెబుతున్నాయి, కానీ ఈ డిజైన్ ఎంపికకు బలమైన ప్రోగ్రామాటిక్ కారణాలు ఉన్నాయి మరియు, ప్రాక్టికల్‌గా, ఈ మార్గం పాండాస్‌కు చాలా సందర్భాల్లో మంచి సమతుల్యతను అందించడానికి సహాయపడుతుంది. అయినప్పటికీ, `None` మరియు `NaN` రెండూ ఉపయోగించడంలో మీరు జాగ్రత్తగా ఉండాల్సిన పరిమితులు కలిగి ఉంటాయి.


### `None`: ఫ్లోట్ కాని లేని డేటా
`None` పాథాన్ నుండి వచ్చినందున, ఇది `'object'` డేటా టైప్ కాని NumPy మరియు pandas అర్రేలలో ఉపయోగించలేము. గుర్తుంచుకోండి, NumPy అర్రేలు (మరియు pandas లోని డేటా నిర్మాణాలు) ఒకే రకమైన డేటాను మాత్రమే కలిగి ఉండగలవు. ఇది పెద్ద స్థాయి డేటా మరియు గణనాత్మక పనులకు వారికి గొప్ప శక్తిని ఇస్తుంది, కానీ ఇది వారి సౌలభ్యాన్ని పరిమితం చేస్తుంది. అలాంటి అర్రేలు "కనిష్ట సాధారణ భాగస్వామి"కి అప్‌కాస్ట్ చేయాలి, అంటే అర్రేలోని అన్ని అంశాలను కవర్ చేసే డేటా టైప్. అర్రేలో `None` ఉన్నప్పుడు, మీరు పాథాన్ ఆబ్జెక్టులతో పని చేస్తున్నారని అర్థం.

దీనిని ప్రదర్శించడానికి, క్రింది ఉదాహరణ అర్రేను పరిగణించండి (దాని `dtype` గమనించండి):


In [9]:
import numpy as np

example1 = np.array([2, None, 6, 8])
example1

array([2, None, 6, 8], dtype=object)

అప్‌కాస్ట్ డేటా రకాల వాస్తవం రెండు పక్క ప్రభావాలను కలిగి ఉంటుంది. మొదటిది, ఆపరేషన్లు కంపైల్ చేయబడిన NumPy కోడ్ కంటే అనువాదం చేయబడిన Python కోడ్ స్థాయిలో నిర్వహించబడతాయి. సారాంశంగా, ఇది `None` ఉన్న `Series` లేదా `DataFrames`తో సంబంధం ఉన్న ఏ ఆపరేషన్లు అయినా నెమ్మదిగా జరుగుతాయని అర్థం. మీరు ఈ పనితీరు తగ్గుదలని గమనించకపోవచ్చు, కానీ పెద్ద డేటాసెట్‌ల కోసం ఇది సమస్యగా మారవచ్చు.

రెండవ పక్క ప్రభావం మొదటి నుండి ఉద్భవిస్తుంది. ఎందుకంటే `None` అసలు `Series` లేదా `DataFrame`లను వనిల్లా Python ప్రపంచంలోకి తిరిగి తీసుకువస్తుంది, అందువల్ల `None` విలువ కలిగిన అర్రేలపై NumPy/pandas సమాహారాలు వంటి `sum()` లేదా `min()` ఉపయోగించడం సాధారణంగా ఒక లోపాన్ని ఉత్పత్తి చేస్తుంది:


In [10]:
example1.sum()

TypeError: ignored

**ప్రధాన విషయం**: పూర్తి సంఖ్యలు మరియు `None` విలువల మధ్య జోడింపు (మరియు ఇతర ఆపరేషన్లు) నిర్వచించబడలేదు, ఇది వాటిని కలిగిన డేటాసెట్‌లతో మీరు చేయగలిగే పనులను పరిమితం చేయవచ్చు.


### `NaN`: లేని ఫ్లోట్ విలువలు

`None` కు విరుద్ధంగా, NumPy (అందువల్ల pandas కూడా) దాని వేగవంతమైన, వెక్టరైజ్డ్ ఆపరేషన్లు మరియు ufuncs కోసం `NaN` ను మద్దతు ఇస్తుంది. చెడు వార్త ఏమిటంటే, `NaN` పై చేసిన ఏ గణిత చర్య అయినా ఎప్పుడూ `NaN` ను ఫలితంగా ఇస్తుంది. ఉదాహరణకు:


In [11]:
np.nan + 1

nan

In [12]:
np.nan * 0

nan

మంచి వార్త: `NaN` ఉన్న అర్రేలపై అమర్చిన సమాహారాలు లోపాలు చూపించవు. చెడు వార్త: ఫలితాలు సమానంగా ఉపయోగకరంగా ఉండవు:


In [13]:
example2 = np.array([2, np.nan, 6, 8]) 
example2.sum(), example2.min(), example2.max()

(nan, nan, nan)

### వ్యాయామం:


In [11]:
# What happens if you add np.nan and None together?


గుర్తుంచుకోండి: `NaN` అనేది కేవలం లేని ఫ్లోటింగ్-పాయింట్ విలువల కోసం మాత్రమే; పూర్తి సంఖ్యలు, స్ట్రింగులు లేదా బూలియన్లకు `NaN` సమానమైనది లేదు.


### `NaN` మరియు `None`: pandas లో null విలువలు

`NaN` మరియు `None` కొంతవరకు వేరుగా ప్రవర్తించగలిగినా, pandas వాటిని పరస్పరం మార్చుకోగలిగేలా నిర్మించబడింది. మనం అర్థం చేసుకోవడానికి, ఒక `Series` యొక్క పూర్తి సంఖ్యలను పరిగణించండి:


In [15]:
int_series = pd.Series([1, 2, 3], dtype=int)
int_series

0    1
1    2
2    3
dtype: int64

### వ్యాయామం:


In [16]:
# Now set an element of int_series equal to None.
# How does that element show up in the Series?
# What is the dtype of the Series?


`Series` మరియు `DataFrame` లలో డేటా సమానత్వాన్ని స్థాపించడానికి డేటా రకాలని అప్‌కాస్టింగ్ చేసే ప్రక్రియలో, pandas `None` మరియు `NaN` మధ్య లోపాల విలువలను సులభంగా మార్చుకుంటుంది. ఈ డిజైన్ లక్షణం కారణంగా, pandas లో `None` మరియు `NaN` ను "null" యొక్క రెండు వేర్వేరు రుచులుగా భావించడం సహాయకరం. నిజానికి, pandas లో లోపాల విలువలతో వ్యవహరించడానికి మీరు ఉపయోగించే కొన్ని ప్రధాన పద్ధతులు వారి పేర్లలో ఈ ఆలోచనను ప్రతిబింబిస్తాయి:

- `isnull()`: లోపాల విలువలను సూచించే బూలియన్ మాస్క్‌ను ఉత్పత్తి చేస్తుంది
- `notnull()`: `isnull()` కు వ్యతిరేకం
- `dropna()`: డేటా యొక్క ఫిల్టర్ చేయబడిన సంస్కరణను తిరిగి ఇస్తుంది
- `fillna()`: లోపాల విలువలు నింపబడిన లేదా అంచనా వేయబడిన డేటా కాపీని తిరిగి ఇస్తుంది

ఇవి నేర్చుకోవడానికి మరియు సౌకర్యవంతంగా ఉపయోగించుకోవడానికి ముఖ్యమైన పద్ధతులు, కాబట్టి వాటిని ప్రతి ఒక్కదానిని కొంత లోతుగా పరిశీలిద్దాం.


### null విలువలను గుర్తించడం

ఇప్పుడు మేము మిస్సింగ్ విలువల ప్రాముఖ్యతను అర్థం చేసుకున్నాము, వాటిని మా డేటాసెట్‌లో గుర్తించాలి, వాటిని నిర్వహించే ముందు.
`isnull()` మరియు `notnull()` రెండూ null డేటాను గుర్తించడానికి మీ ప్రాథమిక పద్ధతులు. ఇవి రెండూ మీ డేటాపై బూలియన్ మాస్క్‌లను తిరిగి ఇస్తాయి.


In [17]:
example3 = pd.Series([0, np.nan, '', None])

In [18]:
example3.isnull()

0    False
1     True
2    False
3     True
dtype: bool

ఫలితాన్ని జాగ్రత్తగా పరిశీలించండి. దాని లో ఏదైనా మీకు ఆశ్చర్యంగా అనిపిస్తుందా? `0` ఒక గణిత శూన్యం అయినప్పటికీ, అది పూర్తిగా సరైన పూర్తి సంఖ్య మరియు pandas దాన్ని అలానే పరిగణిస్తుంది. `''` కొంచెం సున్నితమైనది. మేము దాన్ని సెక్షన్ 1 లో ఖాళీ స్ట్రింగ్ విలువను సూచించడానికి ఉపయోగించినప్పటికీ, అది pandas దృష్టిలో ఒక స్ట్రింగ్ ఆబ్జెక్ట్ మాత్రమే మరియు శూన్యం యొక్క ప్రాతినిధ్యం కాదు.

ఇప్పుడు, దీన్ని తిరగబెట్టుకొని, మీరు ప్రాక్టీస్ లో ఉపయోగించే విధంగా ఈ పద్ధతులను ఉపయోగిద్దాం. మీరు Boolean మాస్క్ లను నేరుగా ``Series`` లేదా ``DataFrame`` సూచికగా ఉపయోగించవచ్చు, ఇది వేరుచేసిన లేని (లేదా ఉన్న) విలువలతో పని చేయడానికి ఉపయోగకరంగా ఉంటుంది.

మేము మొత్తం లేని విలువల సంఖ్య తెలుసుకోవాలనుకుంటే, `isnull()` పద్ధతి ద్వారా ఉత్పత్తి అయిన మాస్క్ పై సుమ్ము చేయవచ్చు.


In [19]:
example3.isnull().sum()

2

### వ్యాయామం:


In [20]:
# Try running example3[example3.notnull()].
# Before you do so, what do you expect to see?


**ప్రధాన విషయం**: మీరు DataFrames లో `isnull()` మరియు `notnull()` పద్ధతులను ఉపయోగించినప్పుడు రెండూ సమాన ఫలితాలను ఇస్తాయి: అవి ఫలితాలను మరియు ఆ ఫలితాల సూచికను చూపిస్తాయి, ఇది మీ డేటాతో పోరాడేటప్పుడు మీకు చాలా సహాయం చేస్తుంది.


### లేని డేటాతో వ్యవహరించడం

> **అధ్యయన లక్ష్యం:** ఈ ఉపవిభాగం ముగిసే వరకు, మీరు DataFrames నుండి నల్ విలువలను ఎప్పుడు మరియు ఎలా మార్చాలి లేదా తొలగించాలి అనేది తెలుసుకోవాలి.

మిషన్ లెర్నింగ్ మోడల్స్ స్వయంగా లేని డేటాతో వ్యవహరించలేవు. కాబట్టి, డేటాను మోడల్‌కు పంపించే ముందు, ఈ లేని విలువలతో వ్యవహరించాలి.

లేని డేటా ఎలా నిర్వహించబడుతుందో దానితో సున్నితమైన వ్యాపారాలు ఉంటాయి, ఇది మీ తుది విశ్లేషణ మరియు వాస్తవ ప్రపంచ ఫలితాలను ప్రభావితం చేయవచ్చు.

లేని డేటాతో వ్యవహరించడానికి ప్రధానంగా రెండు మార్గాలు ఉన్నాయి:


1.   లేని విలువ ఉన్న వరుసను తొలగించండి
2.   లేని విలువను మరొక విలువతో మార్చండి

మేము ఈ రెండు పద్ధతులను మరియు వాటి లాభాలు మరియు నష్టాలను వివరంగా చర్చిస్తాము.


### నల్ విలువలను తొలగించడం

మనం మా మోడల్‌కు పంపే డేటా పరిమాణం దాని పనితీరుపై ప్రత్యక్ష ప్రభావం చూపుతుంది. నల్ విలువలను తొలగించడం అంటే మనం డేటాపాయింట్ల సంఖ్యను తగ్గిస్తున్నాము, అందువల్ల డేటాసెట్ పరిమాణం తగ్గుతుంది. కాబట్టి, డేటాసెట్ చాలా పెద్దదైతే నల్ విలువలతో ఉన్న వరుసలను తొలగించడం మంచిది.

మరొక సందర్భం ఏమిటంటే, ఒక నిర్దిష్ట వరుస లేదా కాలమ్‌లో చాలా మిస్సింగ్ విలువలు ఉండవచ్చు. అప్పుడు, అవి తొలగించబడవచ్చు ఎందుకంటే ఆ వరుస/కాలమ్‌కు సంబంధించిన డేటా ఎక్కువగా లేనందున అవి మన విశ్లేషణకు ఎక్కువ విలువను చేర్చవు.

మిస్సింగ్ విలువలను గుర్తించడం మించి, pandas `Series` మరియు `DataFrame` ల నుండి నల్ విలువలను తొలగించడానికి సౌకర్యవంతమైన మార్గాన్ని అందిస్తుంది. దీన్ని ప్రదర్శించడానికి, మనం `example3` కు తిరిగి వెళ్దాం. `DataFrame.dropna()` ఫంక్షన్ నల్ విలువలతో ఉన్న వరుసలను తొలగించడంలో సహాయపడుతుంది.


In [21]:
example3 = example3.dropna()
example3

0    0
2     
dtype: object

గమనించండి ఇది `example3[example3.notnull()]` నుండి మీ అవుట్పుట్ లాగా కనిపించాలి. ఇక్కడ తేడా ఏమిటంటే, మాస్క్ చేయబడిన విలువలపై మాత్రమే సూచిక వేయడం కాకుండా, `dropna` ఆ లేని విలువలను `Series` `example3` నుండి తొలగించింది.

DataFrames కి రెండు కొలతలు ఉండటం వలన, డేటాను తొలగించడానికి మరిన్ని ఎంపికలు అందుబాటులో ఉంటాయి.


In [22]:
example4 = pd.DataFrame([[1,      np.nan, 7], 
                         [2,      5,      8], 
                         [np.nan, 6,      9]])
example4

0    1  2
0  1.0  NaN  7
1  2.0  5.0  8
2  NaN  6.0  9

(పాండాస్ `NaN` లను అనుకూలించడానికి రెండు కాలమ్స్‌ను ఫ్లోట్స్‌గా అప్‌కాస్ట్ చేస్తున్నట్లు మీరు గమనించారా?)

మీరు ఒకే విలువను `DataFrame` నుండి తొలగించలేరు, కాబట్టి మీరు పూర్తి వరుసలు లేదా కాలమ్స్‌ను తొలగించాలి. మీరు ఏమి చేస్తున్నారో ఆధారంగా, మీరు ఒకటి లేదా మరొకదాన్ని చేయాలనుకోవచ్చు, అందుకే పాండాస్ రెండు కోసం ఎంపికలను ఇస్తుంది. డేటా సైన్స్‌లో, కాలమ్స్ సాధారణంగా వేరియబుల్స్‌ను సూచిస్తాయి మరియు వరుసలు పరిశీలనలను సూచిస్తాయి, కాబట్టి మీరు ఎక్కువగా డేటా వరుసలను తొలగించే అవకాశం ఉంటుంది; `dropna()` యొక్క డిఫాల్ట్ సెట్టింగ్ ఏదైనా నల్ విలువలు ఉన్న అన్ని వరుసలను తొలగించడం:


In [23]:
example4.dropna()

0    1  2
1  2.0  5.0  8

అవసరమైతే, మీరు కాలమ్స్ నుండి NA విలువలను తొలగించవచ్చు. అలా చేయడానికి `axis=1` ఉపయోగించండి:


In [24]:
example4.dropna(axis='columns')

2
0  7
1  8
2  9

గమనించండి, ఇది మీరు ఉంచాలనుకునే చాలా డేటాను డ్రాప్ చేయవచ్చు, ముఖ్యంగా చిన్న డేటాసెట్‌లలో. మీరు కేవలం కొన్ని లేదా అన్ని నల్ విలువలను కలిగిన వరుసలు లేదా కాలమ్స్‌ను మాత్రమే డ్రాప్ చేయాలనుకుంటే ఏమవుతుంది? మీరు `dropna` లో `how` మరియు `thresh` పారామీటర్లతో ఆ సెట్టింగ్స్‌ను నిర్దేశించవచ్చు.

డిఫాల్ట్‌గా, `how='any'` ఉంటుంది (మీరు స్వయంగా తనిఖీ చేయాలనుకుంటే లేదా ఈ మెథడ్‌లో మరే ఇతర పారామీటర్లు ఉన్నాయా చూడాలనుకుంటే, కోడ్ సెల్‌లో `example4.dropna?` ను రన్ చేయండి). మీరు ప్రత్యామ్నాయంగా `how='all'` ను నిర్దేశించవచ్చు, తద్వారా కేవలం అన్ని నల్ విలువలను కలిగిన వరుసలు లేదా కాలమ్స్ మాత్రమే డ్రాప్ చేయబడతాయి. ఈ క్రింది వ్యాయామంలో ఈ చర్యను చూడటానికి మన ఉదాహరణ `DataFrame` ను విస్తరిద్దాం.


In [25]:
example4[3] = np.nan
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

> ముఖ్యాంశాలు:  
1. డేటాసెట్ పెద్దదైతేనే నల్ విలువలను తొలగించడం మంచిది.  
2. ఎక్కువ భాగం డేటా లేని పూర్తి వరుసలు లేదా కాలమ్స్ తొలగించవచ్చు.  
3. `DataFrame.dropna(axis=)` పద్ధతి నల్ విలువలను తొలగించడంలో సహాయపడుతుంది. `axis` ఆర్గ్యుమెంట్ వరుసలు తొలగించాలా లేక కాలమ్స్ తొలగించాలా అనేదాన్ని సూచిస్తుంది.  
4. `how` ఆర్గ్యుమెంట్ కూడా ఉపయోగించవచ్చు. డిఫాల్ట్‌గా ఇది `any` గా సెట్ ఉంటుంది. కాబట్టి, ఇది ఏదైనా నల్ విలువ ఉన్న వరుసలు/కాలమ్స్ మాత్రమే తొలగిస్తుంది. అన్ని విలువలు నల్ ఉన్న వరుసలు/కాలమ్స్ మాత్రమే తొలగించాలంటే దీన్ని `all` గా సెట్ చేయవచ్చు.


### వ్యాయామం:


In [22]:
# How might you go about dropping just column 3?
# Hint: remember that you will need to supply both the axis parameter and the how parameter.


`thresh` పారామీటర్ మీకు మరింత సూక్ష్మ నియంత్రణను ఇస్తుంది: మీరు ఒక వరుస లేదా కాలమ్ నిల్వ చేయబడేందుకు అవసరమైన *నాన్-నల్* విలువల సంఖ్యను సెట్ చేస్తారు:


In [27]:
example4.dropna(axis='rows', thresh=3)

0    1  2   3
1  2.0  5.0  8 NaN

ఇక్కడ, మొదటి మరియు చివరి వరుసలను తొలగించారు, ఎందుకంటే అవి కేవలం రెండు నాన్-నల్ విలువలను మాత్రమే కలిగి ఉన్నాయి.


### నల్ విలువలను నింపడం

కొన్నిసార్లు సరైనవిగా ఉండగల విలువలతో మిస్సింగ్ విలువలను నింపడం అర్థం కలిగిస్తుంది. నల్ విలువలను నింపడానికి కొన్ని సాంకేతికతలు ఉన్నాయి. మొదటిది డొమైన్ జ్ఞానం (డేటాసెట్ ఆధారపడిన విషయం పై జ్ఞానం) ఉపయోగించి మిస్సింగ్ విలువలను ఏదో విధంగా అంచనా వేయడం.

మీరు దీన్ని స్థానంలోనే చేయడానికి `isnull` ఉపయోగించవచ్చు, కానీ మీరు నింపవలసిన విలువలు ఎక్కువగా ఉంటే అది కష్టంగా ఉండవచ్చు. డేటా సైన్స్‌లో ఇది చాలా సాధారణ పని కావడంతో, pandas `fillna`ని అందిస్తుంది, ఇది మిస్సింగ్ విలువలను మీరు ఎంచుకున్న విలువతో మార్చిన `Series` లేదా `DataFrame` యొక్క కాపీని ఇస్తుంది. ఇది ప్రాక్టీస్‌లో ఎలా పనిచేస్తుందో చూడటానికి మరొక ఉదాహరణ `Series`ని సృష్టిద్దాం.


### వర్గీకృత డేటా(సంఖ్యాత్మకంగా లేని)
ముందుగా సంఖ్యాత్మకంగా లేని డేటాను పరిశీలిద్దాం. డేటాసెట్‌లలో, మనకు వర్గీకృత డేటా ఉన్న కాలమ్స్ ఉంటాయి. ఉదా. లింగం, నిజం లేదా తప్పు మొదలైనవి.

ఈ కేసులలో ఎక్కువగా, మిస్సింగ్ విలువలను ఆ కాలమ్ యొక్క `mode` తో భర్తీ చేస్తాము. ఉదాహరణకు, మనకు 100 డేటా పాయింట్లు ఉన్నాయని అనుకుందాం, అందులో 90 మంది నిజం అన్నారు, 8 మంది తప్పు అన్నారు మరియు 2 మంది ఎవరూ పూరించలేదు. అప్పుడు, మొత్తం కాలమ్‌ను పరిగణనలోకి తీసుకుని ఆ 2ని నిజం తో భర్తీ చేయవచ్చు.

మళ్లీ, ఇక్కడ డొమైన్ జ్ఞానం ఉపయోగించవచ్చు. మోడ్ తో భర్తీ చేసే ఉదాహరణను పరిశీలిద్దాం.


In [28]:
fill_with_mode = pd.DataFrame([[1,2,"True"],
                               [3,4,None],
                               [5,6,"False"],
                               [7,8,"True"],
                               [9,10,"True"]])

fill_with_mode

0   1      2
0  1   2   True
1  3   4   None
2  5   6  False
3  7   8   True
4  9  10   True

ఇప్పుడు, ముందుగా మోడ్‌ను కనుగొనుకుందాం, ఆపై `None` విలువను మోడ్‌తో భర్తీ చేద్దాం.


In [29]:
fill_with_mode[2].value_counts()

True     3
False    1
Name: 2, dtype: int64

కాబట్టి, మనం None ను True తో మార్చబోతున్నాము.


In [30]:
fill_with_mode[2].fillna('True',inplace=True)

In [31]:
fill_with_mode

0   1      2
0  1   2   True
1  3   4   True
2  5   6  False
3  7   8   True
4  9  10   True

మనము చూడగలిగినట్లుగా, నల్ విలువ మార్చబడింది. చెప్పాల్సిన అవసరం లేదు, మేము ఎక్కడైనా `'True'` స్థానంలో ఏదైనా రాయవచ్చు మరియు అది మార్చబడేది.


### సంఖ్యా డేటా
ఇప్పుడు, సంఖ్యా డేటాకు వస్తే. ఇక్కడ, మిస్ అయిన విలువలను భర్తీ చేయడానికి రెండు సాధారణ మార్గాలు ఉన్నాయి:

1. వరుస యొక్క మాధ్యమంతో భర్తీ చేయండి
2. వరుస యొక్క సగటుతో భర్తీ చేయండి

మేము మాధ్యమంతో భర్తీ చేస్తాము, అవుట్లయర్లతో కూడిన వక్రీకృత డేటా సందర్భంలో. ఇది ఎందుకంటే మాధ్యమం అవుట్లయర్లకు మన్నించగలదు.

డేటా సాధారణీకరించబడినప్పుడు, మేము సగటును ఉపయోగించవచ్చు, ఎందుకంటే ఆ సందర్భంలో, సగటు మరియు మాధ్యమం చాలా దగ్గరగా ఉంటాయి.

మొదట, మనం సాధారణంగా పంపిణీ చేయబడిన ఒక కాలమ్ తీసుకుందాం మరియు ఆ కాలమ్ యొక్క సగటుతో మిస్ అయిన విలువను నింపుదాం.


In [32]:
fill_with_mean = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [np.nan,4,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  NaN  4  5
3  1.0  6  7
4  2.0  8  9

కాలమ్ యొక్క సగటు విలువ ఏమిటి


In [33]:
np.mean(fill_with_mean[0])

0.0

సగటుతో నింపడం


In [34]:
fill_with_mean[0].fillna(np.mean(fill_with_mean[0]),inplace=True)
fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  0.0  4  5
3  1.0  6  7
4  2.0  8  9

మనము చూడగలిగినట్లుగా, లేని విలువ దాని సగటుతో మార్చబడింది.


ఇప్పుడు మనం మరో డేటాఫ్రేమ్‌ను ప్రయత్నిద్దాం, ఈ సారి None విలువలను కాలమ్ యొక్క మాధ్యమంతో మార్చబోతున్నాము.


In [35]:
fill_with_median = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [0,np.nan,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  NaN  5
3  1  6.0  7
4  2  8.0  9

రెండవ కాలమ్ యొక్క మధ్య విలువ ఏమిటి


In [36]:
fill_with_median[1].median()

4.0

మధ్యమంతో నింపడం


In [37]:
fill_with_median[1].fillna(fill_with_median[1].median(),inplace=True)
fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  4.0  5
3  1  6.0  7
4  2  8.0  9

మనం చూడగలిగినట్లుగా, NaN విలువను కాలమ్ యొక్క మధ్యమ విలువతో మార్చారు.


In [38]:
example5 = pd.Series([1, np.nan, 2, None, 3], index=list('abcde'))
example5

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

మీరు అన్ని నల్ ఎంట్రీలను ఒకే విలువతో పూరించవచ్చు, ఉదాహరణకు `0`:


In [39]:
example5.fillna(0)

a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

> ముఖ్యమైన విషయాలు:
1. డేటా తక్కువగా ఉన్నప్పుడు లేదా లేని డేటాను పూరించడానికి ఒక వ్యూహం ఉన్నప్పుడు మాత్రమే లేని విలువలను పూరించాలి.
2. లేని విలువలను సుమారు అంచనా వేయడానికి డొమైన్ జ్ఞానం ఉపయోగించవచ్చు.
3. వర్గీకరణ డేటా కోసం, ఎక్కువగా లేని విలువలను ఆ కాలమ్ యొక్క మోడ్‌తో భర్తీ చేస్తారు.
4. సంఖ్యాత్మక డేటా కోసం, లేని విలువలను సాధారణంగా సగటు (సాధారణీకరించిన డేటాసెట్‌ల కోసం) లేదా కాలమ్‌ల మధ్య విలువతో పూరిస్తారు.


### వ్యాయామం:


In [40]:
# What happens if you try to fill null values with a string, like ''?


మీరు **ఫార్వర్డ్-ఫిల్** చేయవచ్చు, అంటే చివరి చెల్లుబాటు అయ్యే విలువను ఉపయోగించి నల్ విలువను పూరించవచ్చు:


In [41]:
example5.fillna(method='ffill')

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

మీరు కూడా **back-fill** చేయవచ్చు తద్వారా తదుపరి చెల్లుబాటు అయ్యే విలువను వెనుకకు పంపించి null ని నింపవచ్చు:


In [42]:
example5.fillna(method='bfill')

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

మీరు ఊహించగలిగినట్లుగా, ఇది DataFrames తో కూడా అదే విధంగా పనిచేస్తుంది, కానీ మీరు శూన్య విలువలను నింపడానికి ఒక `axis` ను కూడా నిర్దేశించవచ్చు:


In [43]:
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

In [44]:
example4.fillna(method='ffill', axis=1)

0    1    2    3
0  1.0  1.0  7.0  7.0
1  2.0  5.0  8.0  8.0
2  NaN  6.0  9.0  9.0

గమనించండి, ముందుకు నింపడానికి ముందు విలువ అందుబాటులో లేకపోతే, శూన్య విలువ అలాగే ఉంటుంది.


### వ్యాయామం:


In [45]:
# What output does example4.fillna(method='bfill', axis=1) produce?
# What about example4.fillna(method='ffill') or example4.fillna(method='bfill')?
# Can you think of a longer code snippet to write that can fill all of the null values in example4?


మీరు `fillna` ను ఉపయోగించే విధానంలో సృజనాత్మకంగా ఉండవచ్చు. ఉదాహరణకు, మళ్లీ `example4` ను చూద్దాం, కానీ ఈసారి మిస్సింగ్ విలువలను `DataFrame` లోని అన్ని విలువల సగటుతో నింపుదాం:


In [46]:
example4.fillna(example4.mean())

0    1  2   3
0  1.0  5.5  7 NaN
1  2.0  5.0  8 NaN
2  1.5  6.0  9 NaN

కాలమ్ 3 ఇంకా విలువలేని స్థితిలో ఉందని గమనించండి: డిఫాల్ట్ దిశ విలువలను వరుసల వారీగా నింపడమే.

> **గుర్తుంచుకోవాల్సిన విషయం:** మీ డేటాసెట్‌లలో లేని విలువలతో వ్యవహరించడానికి అనేక మార్గాలు ఉన్నాయి. మీరు ఉపయోగించే నిర్దిష్ట వ్యూహం (వాటిని తొలగించడం, మార్చడం, లేదా మీరు ఎలా మార్చాలో) ఆ డేటా యొక్క ప్రత్యేకతల ఆధారంగా ఉండాలి. మీరు డేటాసెట్‌లను ఎక్కువగా నిర్వహించగలిగే కొద్దీ లేని విలువలతో ఎలా వ్యవహరించాలో మెరుగైన అవగాహన ఏర్పడుతుంది.


### వర్గీకృత డేటాను ఎన్‌కోడ్ చేయడం

యంత్ర అభ్యాస నమూనాలు సంఖ్యలతో మరియు ఏదైనా రకమైన సంఖ్యా డేటాతో మాత్రమే వ్యవహరిస్తాయి. అవి అవును మరియు కాదు మధ్య తేడాను చెప్పలేవు, కానీ 0 మరియు 1 మధ్య తేడాను గుర్తించగలవు. కాబట్టి, మిస్సింగ్ విలువలను పూరించిన తర్వాత, నమూనా అర్థం చేసుకునేందుకు వర్గీకృత డేటాను కొన్ని సంఖ్యా రూపంలో ఎన్‌కోడ్ చేయాలి.

ఎన్‌కోడ్ చేయడం రెండు విధాలుగా చేయవచ్చు. మేము వాటిని తరువాత చర్చించబోతున్నాము.


**లేబుల్ ఎన్‌కోడింగ్**


లేబుల్ ఎన్‌కోడింగ్ అనేది ప్రాథమికంగా ప్రతి వర్గాన్ని ఒక సంఖ్యగా మార్చడం. ఉదాహరణకు, మనకు విమాన ప్రయాణికుల డేటాసెట్ ఉందని మరియు వారి తరగతి ఉన్న ఒక కాలమ్ ['బిజినెస్ క్లాస్', 'ఎకానమీ క్లాస్', 'ఫస్ట్ క్లాస్'] అని తీసుకుందాం. ఈ లేబుల్ ఎన్‌కోడింగ్ చేస్తే, ఇది [0,1,2] గా మారుతుంది. కోడ్ ద్వారా ఒక ఉదాహరణ చూద్దాం. మనం రాబోయే నోట్‌బుక్స్‌లో `scikit-learn` నేర్చుకుంటున్నందున, ఇక్కడ దాన్ని ఉపయోగించము.


In [47]:
label = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
label

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

మొదటి కాలమ్‌పై లేబుల్ ఎంకోడింగ్ చేయడానికి, మేము ప్రతి తరగతిని ఒక సంఖ్యకు మ్యాపింగ్ చేయడం మొదట చేయాలి, ఆ తర్వాత మార్చాలి.


In [48]:
class_labels = {'business class':0,'economy class':1,'first class':2}
label['class'] = label['class'].replace(class_labels)
label

ID  class
0  10      0
1  20      2
2  30      1
3  40      1
4  50      1
5  60      0

మనం చూడగలిగినట్లుగా, అవుట్పుట్ మనం ఊహించినదానికి సరిపోతుంది. కాబట్టి, లేబుల్ ఎంకోడింగ్ ఎప్పుడు ఉపయోగిస్తాము? లేబుల్ ఎంకోడింగ్ క్రింది రెండు సందర్భాలలో ఒకటి లేదా రెండింటిలో ఉపయోగిస్తారు:
1. వర్గాల సంఖ్య ఎక్కువగా ఉన్నప్పుడు
2. వర్గాలు క్రమంలో ఉన్నప్పుడు.


**ఒన్ హాట్ ఎన్‌కోడింగ్**

మరొక రకమైన ఎన్‌కోడింగ్ ఒన్ హాట్ ఎన్‌కోడింగ్. ఈ రకమైన ఎన్‌కోడింగ్‌లో, కాలమ్ యొక్క ప్రతి వర్గం వేరే కాలమ్‌గా జోడించబడుతుంది మరియు ప్రతి డేటాపాయింట్ ఆ వర్గాన్ని కలిగి ఉన్నదో లేదో ఆధారంగా 0 లేదా 1 పొందుతుంది. కాబట్టి, n వేర్వేరు వర్గాలు ఉంటే, n కాలమ్స్ డేటాఫ్రేమ్‌కు జోడించబడతాయి.

ఉదాహరణకు, అదే విమాన తరగతి ఉదాహరణ తీసుకుందాం. వర్గాలు: ['బిజినెస్ క్లాస్', 'ఎకానమీ క్లాస్', 'ఫస్ట్ క్లాస్'] . కాబట్టి, ఒన్ హాట్ ఎన్‌కోడింగ్ చేస్తే, datasetకి క్రింది మూడు కాలమ్స్ జోడించబడతాయి: ['class_business class','class_economy class','class_first class'].


In [49]:
one_hot = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
one_hot

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

మనం మొదటి కాలమ్‌పై వన్ హాట్ ఎంకోడింగ్ చేయండి.


In [50]:
one_hot_data = pd.get_dummies(one_hot,columns=['class'])

In [51]:
one_hot_data

ID  class_business class  class_economy class  class_first class
0  10                     1                    0                  0
1  20                     0                    0                  1
2  30                     0                    1                  0
3  40                     0                    1                  0
4  50                     0                    1                  0
5  60                     1                    0                  0

ప్రతి ఒక హాట్ ఎన్‌కోడ్ చేసిన కాలమ్ 0 లేదా 1 ను కలిగి ఉంటుంది, ఇది ఆ డేటాపాయింట్ కోసం ఆ వర్గం ఉందో లేదో సూచిస్తుంది.


మనం ఎప్పుడు వన్ హాట్ ఎంకోడింగ్ ఉపయోగిస్తాము? వన్ హాట్ ఎంకోడింగ్ క్రింది రెండు సందర్భాలలో ఒకటి లేదా రెండింటిలో ఉపయోగిస్తారు:

1. వర్గాల సంఖ్య మరియు డేటాసెట్ పరిమాణం తక్కువగా ఉన్నప్పుడు.
2. వర్గాలు ఎటువంటి ప్రత్యేక క్రమాన్ని అనుసరించకపోతే.


> ముఖ్యాంశాలు:
1. ఎన్‌కోడింగ్ అనేది సంఖ్యలేని డేటాను సంఖ్యా డేటాగా మార్చడానికి చేయబడుతుంది.
2. ఎన్‌కోడింగ్ రెండు రకాలుగా ఉంటాయి: లేబుల్ ఎన్‌కోడింగ్ మరియు వన్ హాట్ ఎన్‌కోడింగ్, ఇవి డేటాసెట్ అవసరాల ఆధారంగా చేయబడవచ్చు.


## డూప్లికేట్ డేటాను తొలగించడం

> **అధ్యయన లక్ష్యం:** ఈ ఉపవిభాగం ముగిసే వరకు, మీరు DataFrames నుండి డూప్లికేట్ విలువలను గుర్తించి తొలగించడంలో సౌకర్యంగా ఉండాలి.

లేకపోయిన డేటా తో పాటు, మీరు నిజజీవిత డేటాసెట్‌లలో తరచుగా డూప్లికేట్ డేటాను ఎదుర్కొంటారు. అదృష్టవశాత్తు, pandas డూప్లికేట్ ఎంట్రీలను గుర్తించడం మరియు తొలగించడం కోసం సులభమైన మార్గాన్ని అందిస్తుంది.


### డూప్లికేట్లను గుర్తించడం: `duplicated`

మీరు pandas లోని `duplicated` పద్ధతిని ఉపయోగించి డూప్లికేట్ విలువలను సులభంగా గుర్తించవచ్చు, ఇది ఒక `DataFrame` లోని ఎంట్రీ ఒక ముందటి ఎంట్రీ యొక్క డూప్లికేట్ కాదా అనే విషయాన్ని సూచించే బూలియన్ మాస్క్ ను తిరిగి ఇస్తుంది. దీన్ని ప్రదర్శించడానికి మరొక ఉదాహరణ `DataFrame` ను సృష్టిద్దాం.


In [52]:
example6 = pd.DataFrame({'letters': ['A','B'] * 2 + ['B'],
                         'numbers': [1, 2, 1, 3, 3]})
example6

letters  numbers
0       A        1
1       B        2
2       A        1
3       B        3
4       B        3

In [53]:
example6.duplicated()

0    False
1    False
2     True
3    False
4     True
dtype: bool

### డూప్లికేట్లను తొలగించడం: `drop_duplicates`
`drop_duplicates` సాదారణంగా `duplicated` విలువలు అన్నీ `False` అయిన డేటా యొక్క కాపీని తిరిగి ఇస్తుంది:


In [54]:
example6.drop_duplicates()

letters  numbers
0       A        1
1       B        2
3       B        3

`duplicated` మరియు `drop_duplicates` రెండూ డిఫాల్ట్‌గా అన్ని కాలమ్స్‌ను పరిగణలోకి తీసుకుంటాయి కానీ మీరు మీ `DataFrame`లోని కొన్ని కాలమ్స్ మాత్రమే పరిశీలించమని నిర్దేశించవచ్చు:


In [55]:
example6.drop_duplicates(['letters'])

letters  numbers
0       A        1
1       B        2

> **ముఖ్యాంశం:** డూప్లికేట్ డేటాను తొలగించడం ప్రతి డేటా-సైన్స్ ప్రాజెక్ట్‌లో ఒక ముఖ్యమైన భాగం. డూప్లికేట్ డేటా మీ విశ్లేషణల ఫలితాలను మార్చి, మీకు తప్పు ఫలితాలను ఇవ్వవచ్చు!


## వాస్తవ ప్రపంచ డేటా నాణ్యత తనిఖీలు

> **అధ్యయన లక్ష్యం:** ఈ విభాగం ముగింపులో, మీరు అసమానమైన వర్గీకరణ విలువలు, అసాధారణ సంఖ్యా విలువలు (అత్యధికాలు), మరియు వేరియేషన్లతో కూడిన డూప్లికేట్ ఎంటిటీలను గుర్తించి సరిచేయడంలో సౌకర్యంగా ఉండాలి.

తప్పిపోయిన విలువలు మరియు ఖచ్చితమైన డూప్లికేట్లు సాధారణ సమస్యలు అయినప్పటికీ, వాస్తవ ప్రపంచ డేటాసెట్‌లు తరచుగా మరింత సున్నితమైన సమస్యలను కలిగి ఉంటాయి:

1. **అసమానమైన వర్గీకరణ విలువలు**: అదే వర్గం వేరుగా వ్రాయబడటం (ఉదా: "USA", "U.S.A", "United States")
2. **అసాధారణ సంఖ్యా విలువలు**: డేటా ఎంట్రీ లోపాలను సూచించే అత్యధిక అత్యధికాలు (ఉదా: వయస్సు = 999)
3. **సమీప డూప్లికేట్ వరుసలు**: స్వల్ప వేరియేషన్లతో అదే ఎంటిటీని సూచించే రికార్డులు

ఈ సమస్యలను గుర్తించి నిర్వహించడానికి సాంకేతికతలను పరిశీలిద్దాం.


### ఒక నమూనా "అశుభ్రమైన" డేటాసెట్ సృష్టించడం

మొదట, మనం సాధారణంగా నిజమైన ప్రపంచ డేటాలో ఎదురయ్యే సమస్యల రకాలను కలిగిన ఒక నమూనా డేటాసెట్‌ను సృష్టిద్దాం:


In [ ]:
import pandas as pd
import numpy as np

# Create a sample dataset with quality issues
dirty_data = pd.DataFrame({
    'customer_id': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
    'name': ['John Smith', 'Jane Doe', 'John Smith', 'Bob Johnson', 
             'Alice Williams', 'Charlie Brown', 'John  Smith', 'Eva Martinez',
             'Bob Johnson', 'Diana Prince', 'Frank Castle', 'Alice Williams'],
    'age': [25, 32, 25, 45, 28, 199, 25, 31, 45, 27, -5, 28],
    'country': ['USA', 'UK', 'U.S.A', 'Canada', 'USA', 'United Kingdom',
                'United States', 'Mexico', 'canada', 'USA', 'UK', 'usa'],
    'purchase_amount': [100.50, 250.00, 105.00, 320.00, 180.00, 90.00,
                       102.00, 275.00, 325.00, 195.00, 410.00, 185.00]
})

print("Sample 'Dirty' Dataset:")
print(dirty_data)

### 1. అసమంజసమైన వర్గీకరణ విలువలను గుర్తించడం

`country` కాలమ్‌లో ఒకే దేశానికి అనేక ప్రాతినిధ్యాలు ఉన్నాయని గమనించండి. ఈ అసమంజసాలను గుర్తిద్దాం:


In [ ]:
# Check unique values in the country column
print("Unique country values:")
print(dirty_data['country'].unique())
print(f"\nTotal unique values: {dirty_data['country'].nunique()}")

# Count occurrences of each variation
print("\nValue counts:")
print(dirty_data['country'].value_counts())

#### వర్గీకరణ విలువలను ప్రమాణీకరించడం

మేము ఈ విలువలను ప్రమాణీకరించడానికి ఒక మ్యాపింగ్‌ను సృష్టించవచ్చు. ఒక సాదా విధానం అంటే చిన్న అక్షరాలకు మార్చి ఒక మ్యాపింగ్ డిక్షనరీని సృష్టించడం:


In [ ]:
# Create a standardization mapping
country_mapping = {
    'usa': 'USA',
    'u.s.a': 'USA',
    'united states': 'USA',
    'uk': 'UK',
    'united kingdom': 'UK',
    'canada': 'Canada',
    'mexico': 'Mexico'
}

# Standardize the country column
dirty_data['country_clean'] = dirty_data['country'].str.lower().map(country_mapping)

print("Before standardization:")
print(dirty_data['country'].value_counts())
print("\nAfter standardization:")
print(dirty_data[['country_clean']].value_counts())

**వికల్పం: ఫజ్జీ మ్యాచ్ చేయడం ఉపయోగించడం**

మరింత క్లిష్టమైన సందర్భాల కోసం, మనం `rapidfuzz` లైబ్రరీతో ఫజ్జీ స్ట్రింగ్ మ్యాచ్ చేయడం ఉపయోగించి సాదృశ్యమైన స్ట్రింగులను ఆటోమేటిక్‌గా గుర్తించవచ్చు:


In [ ]:
try:
    from rapidfuzz import process, fuzz
except ImportError:
    print("rapidfuzz is not installed. Please install it with 'pip install rapidfuzz' to use fuzzy matching.")
    process = None
    fuzz = None

# Get unique countries
unique_countries = dirty_data['country'].unique()

# For each country, find similar matches
if process is not None and fuzz is not None:
    print("Finding similar country names (similarity > 70%):")
    for country in unique_countries:
        matches = process.extract(country, unique_countries, scorer=fuzz.ratio, limit=3)
        # Filter matches with similarity > 70 and not identical
        similar = [m for m in matches if m[1] > 70 and m[0] != country]
        if similar:
            print(f"\n'{country}' is similar to:")
            for match, score, _ in similar:
                print(f"  - '{match}' (similarity: {score}%)")
else:
    print("Skipping fuzzy matching because rapidfuzz is not available.")

### 2. అసాధారణ సంఖ్యా విలువలను గుర్తించడం (అవుట్లయర్స్)

`age` కాలమ్‌ను చూస్తే, 199 మరియు -5 వంటి కొన్ని అనుమానాస్పద విలువలు ఉన్నాయి. ఈ అవుట్లయర్స్‌ను గుర్తించడానికి గణాంక పద్ధతులను ఉపయోగిద్దాం.


In [ ]:
# Display basic statistics
print("Age column statistics:")
print(dirty_data['age'].describe())

# Identify impossible values using domain knowledge
print("\nRows with impossible age values (< 0 or > 120):")
impossible_ages = dirty_data[(dirty_data['age'] < 0) | (dirty_data['age'] > 120)]
print(impossible_ages[['customer_id', 'name', 'age']])

#### IQR (ఇంటర్కార్టైల్ రేంజ్) పద్ధతి ఉపయోగించడం

IQR పద్ధతి అత్యంత విలువలకు తక్కువ సున్నితమైన, అవుట్లయర్ గుర్తింపు కోసం బలమైన గణాంక పద్ధతి.


In [ ]:
# Calculate IQR for age (excluding impossible values)
valid_ages = dirty_data[(dirty_data['age'] >= 0) & (dirty_data['age'] <= 120)]['age']

Q1 = valid_ages.quantile(0.25)
Q3 = valid_ages.quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print(f"IQR-based outlier bounds for age: [{lower_bound:.2f}, {upper_bound:.2f}]")

# Identify outliers
age_outliers = dirty_data[(dirty_data['age'] < lower_bound) | (dirty_data['age'] > upper_bound)]
print(f"\nRows with age outliers:")
print(age_outliers[['customer_id', 'name', 'age']])

#### Z-స్కోర్ పద్ధతి ఉపయోగించడం

Z-స్కోర్ పద్ధతి సగటు నుండి స్టాండర్డ్ డివియేషన్ల ఆధారంగా అవుట్‌లయర్లను గుర్తిస్తుంది:


In [ ]:
try:
    from scipy import stats
except ImportError:
    print("scipy is required for Z-score calculation. Please install it with 'pip install scipy' and rerun this cell.")
else:
    # Calculate Z-scores for age, handling NaN values
    age_nonan = dirty_data['age'].dropna()
    zscores = np.abs(stats.zscore(age_nonan))
    dirty_data['age_zscore'] = np.nan
    dirty_data.loc[age_nonan.index, 'age_zscore'] = zscores

    # Typically, Z-score > 3 indicates an outlier
    print("Rows with age Z-score > 3:")
    zscore_outliers = dirty_data[dirty_data['age_zscore'] > 3]
    print(zscore_outliers[['customer_id', 'name', 'age', 'age_zscore']])

    # Clean up the temporary column
    dirty_data = dirty_data.drop('age_zscore', axis=1)

#### అవుట్లయర్స్‌ను నిర్వహించడం

ఒకసారి గుర్తించిన తర్వాత, అవుట్లయర్స్‌ను అనేక విధాలుగా నిర్వహించవచ్చు:
1. **తొలగించు**: అవుట్లయర్స్ ఉన్న వరుసలను తొలగించండి (వీటిని తప్పులుగా ఉంటే)
2. **క్యాప్ చేయండి**: సరిహద్దు విలువలతో మార్చండి
3. **NaN తో మార్చండి**: లేని డేటాగా పరిగణించి ఇంప్యూటేషన్ సాంకేతికతలను ఉపయోగించండి
4. **ఉంచండి**: అవి చట్టబద్ధమైన అత్యంత విలువలు అయితే


In [ ]:
# Create a cleaned version by replacing impossible ages with NaN
dirty_data['age_clean'] = dirty_data['age'].apply(
    lambda x: np.nan if (x < 0 or x > 120) else x
)

print("Age column before and after cleaning:")
print(dirty_data[['customer_id', 'name', 'age', 'age_clean']])

### 3. సమీప-నకిలీ వరుసలను గుర్తించడం

మన డేటాసెట్‌లో "జాన్ స్మిత్" కోసం కొంత భిన్నమైన విలువలతో అనేక ఎంట్రీలు ఉన్నాయని గమనించండి. పేరుల సమానత ఆధారంగా సంభావ్య నకిలీలను గుర్తిద్దాం.


In [ ]:
# First, let's look at exact name matches (ignoring extra whitespace)
dirty_data['name_normalized'] = dirty_data['name'].str.strip().str.lower()

print("Checking for duplicate names:")
duplicate_names = dirty_data[dirty_data.duplicated(['name_normalized'], keep=False)]
print(duplicate_names.sort_values('name_normalized')[['customer_id', 'name', 'age', 'country']])

#### ఫజీ మ్యాచ్ చేయడం ద్వారా సమీప-నకిలీలను కనుగొనడం

మరింత సున్నితమైన నకిలీ గుర్తింపు కోసం, సమానమైన పేర్లను కనుగొనడానికి ఫజీ మ్యాచ్ చేయడాన్ని ఉపయోగించవచ్చు:


In [ ]:
try:
    from rapidfuzz import process, fuzz

    # Function to find potential duplicates
    def find_near_duplicates(df, column, threshold=90):
        """
        Find near-duplicate entries in a column using fuzzy matching.
        
        Parameters:
        - df: DataFrame
        - column: Column name to check for duplicates
        - threshold: Similarity threshold (0-100)
        
        Returns: List of potential duplicate groups
        """
        values = df[column].unique()
        duplicate_groups = []
        checked = set()
        
        for value in values:
            if value in checked:
                continue
                
            # Find similar values
            matches = process.extract(value, values, scorer=fuzz.ratio, limit=len(values))
            similar = [m[0] for m in matches if m[1] >= threshold]
            
            if len(similar) > 1:
                duplicate_groups.append(similar)
                checked.update(similar)
        
        return duplicate_groups

    # Find near-duplicate names
    duplicate_groups = find_near_duplicates(dirty_data, 'name', threshold=90)

    print("Potential duplicate groups:")
    for i, group in enumerate(duplicate_groups, 1):
        print(f"\nGroup {i}:")
        for name in group:
            matching_rows = dirty_data[dirty_data['name'] == name]
            print(f"  '{name}': {len(matching_rows)} occurrence(s)")
            for _, row in matching_rows.iterrows():
                print(f"    - Customer {row['customer_id']}: age={row['age']}, country={row['country']}")
except ImportError:
    print("rapidfuzz is not installed. Skipping fuzzy matching for near-duplicates.")

#### డూప్లికేట్లను నిర్వహించడం

ఒకసారి గుర్తించిన తర్వాత, మీరు డూప్లికేట్లను ఎలా నిర్వహించాలో నిర్ణయించుకోవాలి:
1. **మొదటి సంభవాన్ని ఉంచండి**: `drop_duplicates(keep='first')` ఉపయోగించండి
2. **చివరి సంభవాన్ని ఉంచండి**: `drop_duplicates(keep='last')` ఉపయోగించండి
3. **సమాచారాన్ని సమ్మేళనం చేయండి**: డూప్లికేట్ వరుసల నుండి సమాచారాన్ని కలపండి
4. **మాన్యువల్ సమీక్ష**: మానవ సమీక్ష కోసం ఫ్లాగ్ చేయండి


In [ ]:
# Example: Remove duplicates based on normalized name, keeping first occurrence
cleaned_data = dirty_data.drop_duplicates(subset=['name_normalized'], keep='first')

print(f"Original dataset: {len(dirty_data)} rows")
print(f"After removing name duplicates: {len(cleaned_data)} rows")
print(f"Removed: {len(dirty_data) - len(cleaned_data)} duplicate rows")

print("\nCleaned dataset:")
print(cleaned_data[['customer_id', 'name', 'age', 'country_clean']])

### సారాంశం: పూర్తి డేటా శుభ్రపరిచే పైప్‌లైన్

మనం అన్నింటినీ ఒక సమగ్ర శుభ్రపరిచే పైప్‌లైన్‌లో కలపుదాం:


In [ ]:
def clean_dataset(df):
    """
    Comprehensive data cleaning function.
    """
    # Create a copy to avoid modifying the original
    cleaned = df.copy()
    
    # 1. Standardize categorical values (country)
    country_mapping = {
        'usa': 'USA', 'u.s.a': 'USA', 'united states': 'USA',
        'uk': 'UK', 'united kingdom': 'UK',
        'canada': 'Canada', 'mexico': 'Mexico'
    }
    cleaned['country'] = cleaned['country'].str.lower().map(country_mapping)
    
    # 2. Clean abnormal age values
    cleaned['age'] = cleaned['age'].apply(
        lambda x: np.nan if (x < 0 or x > 120) else x
    )
    
    # 3. Remove near-duplicate names (normalize whitespace)
    cleaned['name'] = cleaned['name'].str.strip()
    cleaned = cleaned.drop_duplicates(subset=['name'], keep='first')
    
    return cleaned

# Apply the cleaning pipeline
final_cleaned_data = clean_dataset(dirty_data)

print("Before cleaning:")
print(f"  Rows: {len(dirty_data)}")
print(f"  Unique countries: {dirty_data['country'].nunique()}")
print(f"  Invalid ages: {((dirty_data['age'] < 0) | (dirty_data['age'] > 120)).sum()}")

print("\nAfter cleaning:")
print(f"  Rows: {len(final_cleaned_data)}")
print(f"  Unique countries: {final_cleaned_data['country'].nunique()}")
print(f"  Invalid ages: {((final_cleaned_data['age'] < 0) | (final_cleaned_data['age'] > 120)).sum()}")

print("\nCleaned dataset:")
print(final_cleaned_data[['customer_id', 'name', 'age', 'country', 'purchase_amount']])

### 🎯 సవాలు వ్యాయామం

ఇప్పుడు మీ వారం! క్రింద ఒక కొత్త డేటా వరుస ఉంది, ఇందులో అనేక నాణ్యత సమస్యలు ఉన్నాయి. మీరు చేయగలరా:

1. ఈ వరుసలోని అన్ని సమస్యలను గుర్తించండి
2. ప్రతి సమస్యను శుభ్రం చేయడానికి కోడ్ రాయండి
3. శుభ్రం చేసిన వరుసను డేటాసెట్‌లో జోడించండి

ఇది సమస్యాత్మక డేటా:


In [ ]:
# New problematic row
new_row = pd.DataFrame({
    'customer_id': [13],
    'name': ['  Diana  Prince  '],  # Extra whitespace
    'age': [250],  # Impossible age
    'country': ['U.S.A.'],  # Inconsistent format
    'purchase_amount': [150.00]
})

print("New row to clean:")
print(new_row)

# TODO: Your code here to clean this row
# Hints:
# 1. Strip whitespace from the name
# 2. Check if the name is a duplicate (Diana Prince already exists)
# 3. Handle the impossible age value
# 4. Standardize the country name

# Example solution (uncomment and modify as needed):
# new_row_cleaned = new_row.copy()
# new_row_cleaned['name'] = new_row_cleaned['name'].str.strip()
# new_row_cleaned['age'] = np.nan  # Invalid age
# new_row_cleaned['country'] = 'USA'  # Standardized
# print("\nCleaned row:")
# print(new_row_cleaned)

### ముఖ్యమైన విషయాలు

1. **అసమానమైన వర్గాలు** వాస్తవ ప్రపంచ డేటాలో సాధారణం. ఎప్పుడూ ప్రత్యేక విలువలను తనిఖీ చేసి, మ్యాపింగ్‌లు లేదా ఫజీ మ్యాచ్ చేయడం ద్వారా వాటిని ప్రమాణీకరించండి.

2. **అత్యంత విలక్షణాలు** మీ విశ్లేషణపై గణనీయంగా ప్రభావం చూపవచ్చు. వాటిని గుర్తించడానికి డొమైన్ జ్ఞానం మరియు గణాంక పద్ధతులు (IQR, Z-స్కోర్) ఉపయోగించండి.

3. **సమీప-నకిలీలు** ఖచ్చితమైన నకిలీల కంటే గుర్తించడం కష్టం. వాటిని గుర్తించడానికి ఫజీ మ్యాచ్ చేయడం మరియు డేటాను సాధారణీకరించడం (తక్కువ అక్షరాలు, ఖాళీలను తొలగించడం) పరిగణించండి.

4. **డేటా శుభ్రపరిచే ప్రక్రియ పునరావృతం**. మీరు అనేక సాంకేతికతలను వరుసగా ఉపయోగించి, ఫలితాలను సమీక్షించి, చివరికి శుభ్రపరిచిన డేటాసెట్‌ను నిర్ధారించాలి.

5. **మీ నిర్ణయాలను డాక్యుమెంట్ చేయండి**. మీరు ఏ శుభ్రపరిచే దశలను ఎందుకు ఉపయోగించారో ట్రాక్ చేయండి, ఇది పునరుత్పత్తి మరియు పారదర్శకతకు ముఖ్యమైనది.

> **ఉత్తమ ఆచారం:** ఎప్పుడూ మీ అసలు "కురుపు" డేటా కాపీని ఉంచండి. మీ మూల డేటా ఫైళ్లను ఎప్పుడూ మళ్లీ రాయవద్దు - స్పష్టమైన పేర్లతో శుభ్రపరిచిన వెర్షన్లను సృష్టించండి, ఉదాహరణకు `data_cleaned.csv`.


---

<!-- CO-OP TRANSLATOR DISCLAIMER START -->
**అస్పష్టత**:  
ఈ పత్రాన్ని AI అనువాద సేవ [Co-op Translator](https://github.com/Azure/co-op-translator) ఉపయోగించి అనువదించబడింది. మేము ఖచ్చితత్వానికి ప్రయత్నించినప్పటికీ, ఆటోమేటెడ్ అనువాదాల్లో పొరపాట్లు లేదా తప్పిదాలు ఉండవచ్చు. మూల పత్రం దాని స్వదేశీ భాషలో అధికారిక మూలంగా పరిగణించాలి. ముఖ్యమైన సమాచారానికి, ప్రొఫెషనల్ మానవ అనువాదం సిఫార్సు చేయబడుతుంది. ఈ అనువాదం వాడకంలో ఏర్పడిన ఏవైనా అపార్థాలు లేదా తప్పుదారితీసే అర్థాలు కోసం మేము బాధ్యత వహించము.
<!-- CO-OP TRANSLATOR DISCLAIMER END -->
